### ISSUE TITLE
Close tabs by middle clicking

### ISSUE THREAD

**Issue by AUTplayed** [onivim/oni#2069]
Would love to be able to close editor tabs by clicking mouse3 (middle click, mouse wheel click...) like in many other programs.

**Comment by texhnolyze** [comment]
I am currently having a look at this and did a first shot implementation of this, by adding a mousedown listener to both the tabs name and file icon (similar to the click listener to switch tabs).

Check https://github.com/onivim/oni/compare/master...texhnolyze:middle-click-tab-close?expand=1

I am however not very happy with this solution as it stands and would appreciate some suggestions:
* One option would be to combine the onClick and onMouseDown event handlers and select/close tabs based on the ```React.MouseEvent.button``` value (which I already do for the middle click)
* Another options if something like that is achievable nicely with react (I come more from an angular background) would be to write a general purpose middle click event handler to run any given function on a middle click of an element e.g. 
```Javascript
<div  className="name"
         onClick={this.props.onClickName}
         onMiddleClick={this.props.onMiddleClick}>
```

On another note I noticed, that tabs do not close if you only have a single one which suprised me but makes sense from a neovim perspective. Should I track that in an issue?

**Comment by Akin909** [comment]
@texhnolyze I think from a react perspective option one if *much* more preferable to option two especially as we have a few issues already which are caused by very general event handlers occasionally swallowing or obscuring events. I think having an on click that also listens for a middle click would be a good solution

Tbh never used middle click or listened for it so can't advise much there re the actual event, re the tabs not closing if you only have a single one the reason for that is that an issue was raised a while back that if oni had only one tab remaining and it was closed neovim/oni would close which is how neovim responds to that so what is done instead now is a no name buffer is created on purpose to prevent that behaviour

**Comment by texhnolyze** [comment]
@Akin909 I did an implementation of an ```onMouseDown``` event handler to handle both the click and middle click as sadly ```onClick``` only gets triggered when using left mouse button. So I have to check the events button value for left and middle click to prevent unexpected behaviour on right clicks

Concerning the tabs from what I gather from your response is, that when I have a single tab of let's say ```resume.md``` and I close it, it should be closed but a new empty buffer with the oni tab name ```[No Name]``` should be opened? Did I get that correctly? 
Because what is happening for me is that, when I try to close ```resume.md``` there is just nothing happening, which led me to assume I did not click the close button correctly and clicking it again the first time I encountered it.

Also concerning testing, I tried setting up a ```<Tabs />``` ui-test, which actually tests clicks and middle clicks, but have some issues with it. When trying to mount a ```<Tabs>``` element in a test (so that the child elements are fully rendered as well), what happens is that I run into issues due to the ```<FileIcon>``` element in the tab which uses icon themes configured under ```"ui.iconTheme``` and is thus asynchronous, leading to me trying to access members on ```_icons```, which is null. (see: https://github.com/texhnolyze/oni/blob/b96cbc41319ba4376690f2fb7031ffddb1d7d541/browser/src/UI/components/Tabs.tsx#L172,176) 
Any suggestions on how to solve this testing issue?
